clippy_lints/src/methods/mod.rs

```diff
@@ -1189,93 +1167,115 @@ fn lint_expect_fun_call(cx: &LateContext<'_, '_>, expr: &hir::Expr, method_span:
-            let mut args = vec![snippet(cx, fmt_spec.span, "..").into_owned()];
+    let span_replace_word = method_span.with_hi(expr.span.hi());
 
-            args.extend(generate_format_arg_snippet(cx, fmt_args, &mut applicability));
+    let mut applicability = Applicability::MachineApplicable;
 
-            let sugg = args.join(", ");
+    //Special handling for `format!` as arg_root
+    if let hir::ExprKind::Call(ref inner_fun, ref inner_args) = arg_root.node {
+        if is_expn_of(inner_fun.span, "format").is_some() && inner_args.len() == 1 {
+            if let hir::ExprKind::Call(_, format_args) = &inner_args[0].node {
+                let fmt_spec = &format_args[0];
+                let fmt_args = &format_args[1];
 
-            span_lint_and_sugg(
-                cx,
-                EXPECT_FUN_CALL,
-                span_replace_word,
-                &format!("use of `{}` followed by a function call", name),
-                "try this",
-                format!("unwrap_or_else({} panic!({}))", closure, sugg),
-                applicability,
-            );
+                let mut applicability = Applicability::MachineApplicable;
+                let mut args = vec![snippet(cx, fmt_spec.span, "..").into_owned()];
 
-            return;
-        }
+                args.extend(generate_format_arg_snippet(cx, fmt_args, &mut applicability));
 
-        let mut applicability = Applicability::MachineApplicable;
-        let sugg: Cow<'_, _> = snippet_with_applicability(cx, arg.span, "..", &mut applicability);
+                let sugg = args.join(", ");
 
-        span_lint_and_sugg(
-            cx,
-            EXPECT_FUN_CALL,
-            span_replace_word,
-            &format!("use of `{}` followed by a function call", name),
-            "try this",
-            format!("unwrap_or_else({} {{ let msg = {}; panic!(msg) }}))", closure, sugg),
-            applicability,
-        );
+                span_lint_and_sugg(
+                    cx,
+                    EXPECT_FUN_CALL,
+                    span_replace_word,
+                    &format!("use of `{}` followed by a function call", name),
+                    "try this",
+                    format!("unwrap_or_else({} panic!({}))", closure_args, sugg),
+                    applicability,
+                );
+
+                return;
+            }
+        }
     }
 
-    if args.len() == 2 {
-        match args[1].node {
-            hir::ExprKind::Lit(_) => {},
-            _ => check_general_case(cx, name, method_span, &args[0], &args[1], expr.span),
+    // If root_arg is `&'static str` or `String` we can use it directly in the `panic!` call otherwise
+    // we must use `to_string` to convert it.
+    let mut arg_root_snippet: Cow<'_, _> = snippet_with_applicability(cx, arg_root.span, "..", &mut applicability);
+    let arg_root_ty = cx.tables.expr_ty(arg_root);
+    let mut requires_conv = !match_type(cx, arg_root_ty, &paths::STRING);
```

Comment:
Also this `requires_conv` and the 4 lines below would benefit from being in a separate function with a small comment.